wallet-core-travel/src/utils/UserOp.ts

```diff
@@ -0,0 +1,113 @@
+/**
+ * fork from:
+ * @link https://github.com/eth-infinitism/account-abstraction/blob/develop/test/UserOp.ts
+ */
+
+import { arrayify, defaultAbiCoder, keccak256 } from 'ethers/lib/utils'
+import { Wallet } from 'ethers'
+import { ecsign, toRpcSig, keccak256 as keccak256_buffer } from 'ethereumjs-util'
+import { UserOperation } from '../entity/userOperation'
+import { EntryPointAddress } from '../defines';
+
+function encode(typevalues: Array<{ type: string, val: any }>, forSignature: boolean): string {
+  const types = typevalues.map(typevalue => typevalue.type === 'bytes' && forSignature ? 'bytes32' : typevalue.type)
+  const values = typevalues.map((typevalue) => typevalue.type === 'bytes' && forSignature ? keccak256(typevalue.val) : typevalue.val)
+  return defaultAbiCoder.encode(types, values)
+}
+
+export function packUserOp(op: UserOperation, forSignature = true): string {
+  if (forSignature) {
+    // lighter signature scheme (must match UserOperation#pack): do encode a zero-length signature, but strip afterwards the appended zero-length value
+    const userOpType = {
+      components: [
+        { type: 'address', name: 'sender' },
+        { type: 'uint256', name: 'nonce' },
+        { type: 'bytes', name: 'initCode' },
+        { type: 'bytes', name: 'callData' },
+        { type: 'uint256', name: 'callGas' },
+        { type: 'uint256', name: 'verificationGas' },
+        { type: 'uint256', name: 'preVerificationGas' },
+        { type: 'uint256', name: 'maxFeePerGas' },
+        { type: 'uint256', name: 'maxPriorityFeePerGas' },
+        { type: 'address', name: 'paymaster' },
+        { type: 'bytes', name: 'paymasterData' },
+        { type: 'bytes', name: 'signature' }
+      ],
+      name: 'userOp',
+      type: 'tuple'
+    }
+    let encoded = defaultAbiCoder.encode([userOpType as any], [{ ...op, signature: '0x' }])
```

Comment:
I need some background to know this sentence....
for ex, .encode is a ENUM list? how can we select a special algorithm?

Reply:
and how about the signature be produced?